File: src/main/java/ru/isb/bot/services/MessageServiceImpl.java
package ru.isb.bot.services;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.reflect.TypeToken;
import lombok.RequiredArgsConstructor;
import okhttp3.Response;
import org.springframework.stereotype.Service;
import ru.isb.bot.clients.ChatGPTClient;
import ru.isb.bot.clients.NextcloudClient;
import ru.isb.bot.clients.StudyScheduleClient;
import ru.isb.bot.dto.ScheduleDTO;
import ru.isb.bot.utils.MessageUtils;
import ru.isb.bot.utils.StringUtils;

import java.io.*;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

import static ru.isb.bot.clients.StudyScheduleClientImpl.URL;

@Service
@RequiredArgsConstructor
public class MessageServiceImpl implements MessageService {

    private final StudyScheduleClient client;
    private final ChatGPTClient gptClient;
    private final NextcloudClient nextcloudClient;

    public static final String MESSAGE_GPT_SPLIT = "GPT -> ";

    private final List<String> list = Arrays.asList("Андреев Р.Н.",
            "Буркова П.В.",
            "Кетов А.А.",
            "Колышкин Ю.А.",
            "Кузаев Р.Р.",
            "Маркин К.А.",
            "Сидякин И.О.",
            "Тебеньков В.А.",
            "Забрудский А.В.");

    @Override
    public String getSchedulesWeek() throws IOException, InterruptedException {
        LocalDate date = LocalDate.now();
        LocalDate endDate = LocalDate.now().plusDays(10);
        Response response = client.getTimetableOfClasses(date, endDate);
        if (response.isSuccessful()) {
            if (response.body() != null) {
                Gson gson = new GsonBuilder()
                        .setPrettyPrinting()
                        .create();
                List<ScheduleDTO> schedules = gson.fromJson(response.body().string(), new TypeToken<ArrayList<ScheduleDTO>>() {}.getType());
                return MessageUtils.formatMessage(schedules);
            }
        } else {
            return "<b>Ошибка на стороне сервера ПГАТУ!</b>" +
                    "\nURL: " + URL +
                    "\nКод ошибки: " + response.code() +
                    (response.body() != null ? "\nТекст ошибки: " + response.body().string() : null) +
                    "\n<b>Необходимо написать Мелехину Максиму Игоревичу на учебном портале.</b>";
        }

        return null;
    }

    @Override
    public String getListGroup() {
        list.sort(String::compareTo);
        return "<b>Список группы: </b>\n" + IntStream.range(0, list.size())
                .mapToObj(i -> i + 1 + ") " + list.get(i) + "\n").collect(Collectors.joining());
    }

    @Override
    public String getAnswerMessage(String text) {
        return StringUtils.replaceHTML(gptClient.getAnswerGPT(text.split(MESSAGE_GPT_SPLIT)[1]));
    }

    @Override
    public void sendFileToNextcloud(File file, String fileName) {
        nextcloudClient.uploadFile(file, fileName);
    }
}

File: src/main/java/ru/isb/bot/services/MessageServiceImpl.kt
package ru.isb.bot.services

import com.google.gson.GsonBuilder
import com.google.gson.JsonSyntaxException
import com.google.gson.reflect.TypeToken
import org.apache.logging.log4j.kotlin.Logging
import org.springframework.stereotype.Service
import ru.isb.bot.clients.ChatGPTClient
import ru.isb.bot.clients.NextcloudClient
import ru.isb.bot.clients.StudyScheduleClient
import ru.isb.bot.clients.StudyScheduleClientImpl
import ru.isb.bot.dto.ScheduleDTO
import ru.isb.bot.utils.MessageUtils.Companion.formatMessage
import java.io.File
import java.io.IOException
import java.time.LocalDate
import java.util.stream.Collectors
import java.util.stream.IntStream

@Service
class MessageServiceImpl(
    private val studyScheduleClient: StudyScheduleClient,
    private val gptClient: ChatGPTClient,
    private val nextcloudClient: NextcloudClient
) : MessageService, Logging {

    companion object {
        val MESSAGE_GPT_SPLIT = "GPT -> "
    }

    private val list: List<String> = mutableListOf(
        "Андреев Р.Н.",
        "Буркова П.В.",
        "Кетов А.А.",
        "Колышкин Ю.А.",
        "Кузаев Р.Р.",
        "Маркин К.А.",
        "Сидякин И.О.",
        "Тебеньков В.А.",
        "Забрудский А.В."
    )

    @Throws(IOException::class, InterruptedException::class)
    override fun getSchedulesWeek(): String {
        val date = LocalDate.now()
        val endDate = LocalDate.now().plusDays(10)
        val response = studyScheduleClient.getTimetableOfClasses(date, endDate)
        if (response.isSuccessful) {
            val gson = GsonBuilder()
                .setPrettyPrinting()
                .create()

            return try {
                val schedules = gson.fromJson<List<ScheduleDTO>>(response.body?.string(),
                    object : TypeToken<List<ScheduleDTO>>() {}.type
                )

                formatMessage(schedules)
            } catch (e: JsonSyntaxException) {
                "Расписание нет! Разработчик не виновен, бейте палкой того кто составляет расписание :D"
            }

        } else {
            return """
                     <b>Ошибка на стороне сервера ПГАТУ!</b>
                     URL: ${StudyScheduleClientImpl.URL}
                     Код ошибки: ${response.code}
                 """.trimIndent() +
                    (if (response.body != null) """
                                             Текст ошибки: ${response.body?.string()}
                                            """.trimIndent()
                    else null) +
                    "\n<b>Необходимо написать Мелехину Максиму Игоревичу на учебном портале.</b>"
        }
    }

    override fun getListGroup(): String {
        list.sorted();
        return "<b>Список группы: </b>\n" +
                IntStream.range(0, list.size)
                    .mapToObj { i: Int ->
                        (i + 1).toString() + ") " + list[i] + "\n"
                    }.collect(Collectors.joining());
    }

    override fun getAnswerGPTMessage(text: String): String {
        return gptClient.getAnswerGPT(
                text.split(MESSAGE_GPT_SPLIT)[1]
            )
    }

    override fun sendFileToNextcloud(file: File, fileName: String) {
        nextcloudClient.uploadFile(file, fileName)
    }
}